### ISSUE TITLE
Linux x86 builds (or documentation for it)

### ISSUE THREAD

**Issue by weltkante** [dotnet/source-build#1235]
It's been a longstanding issue that there are no x86 builds for linux, even though it technically is already (supposedly) possible, closed issues in corefx and coreclr exist with people reporting they have build it - they even have documents how to cross compile - unfortunately nobody ever bothered to document how this works on a big scale, how to produce a full build usable for creating x86 applications.

I've been pointed by @jkotas from dotnet/coreclr#9265 to this repo since it integrates the build of the individual repos, but haven't been able to sucessfully build anything yet.

What I have tried before source-build:
- I'm using Ubuntu 16.04 x86_64 in a VM
- I'm able to build corefx and coreclr individually following their documentation
- I can cross-compile coreclr for x86 following their documentation
- I fail to create a custom corefx build because there is no documentation how to feed it the custom coreclr build
- I didn't even attempt to build anything past corefx yet

What I've tried with source-build:
- the only document talking even remotely about the topic of creating a build for a non-official RID is the [bootstrap documentation](https://github.com/dotnet/source-build/blob/release/3.0/Documentation/boostrap-new-os.md)
- the bootstrap documentation is not easy to follow for the usecase of creating a Linux x86 build, in particular setting up a working seedcli took a lot of trial and error
  * the document says it expects the seedcli to use the same architecture as the target build, so I tried using a win-x86 one but that didn't work out
  * then I read the whole document and realized at the end it explained how to build for different architectures, so I tried using linux-x64 as seedcli; this failed as well when following the recommendation of picking the version from [DotnetCLIVersion.txt](https://github.com/dotnet/coreclr/blob/master/DotnetCLIVersion.txt) - it complains about missing NuGetFallbackFolder, even after running `dotnet new console` with the seedcli (which fills the NuGetFallbackFolder) it has the same error
  * ignoring the recommendation and just picking a 3.0 preview7 sdk as seedcli allows to get a bit further, now it starts compiling things but fails with coreclr, missing the `--cross` flag

I could probably try to create a plain x86 VM instead of x86_64 but I think I have tried that before and was running out of memory in some build step. It's been a long time, maybe it works better now, or maybe I remember wrong.

Anyways, my request is, if there's not going to be official Linux x86 builds it would be great if it would be at least possible to do custom builds - or if its too complicated to integrate into the tooling, at least document how to do them. I guess _someone_ must have tried it?

(Background: x64 builds of .NET Core require 2x the memory so they require paying for additional RAM when deploying .NET Core applications instead of Mono applications. This has been a blocker on fully migrating to .NET Core for years.)

**Comment by omajid** [comment]
> Background: x64 builds of .NET Core require 2x the memory so they require paying for additional RAM when deploying .NET Core applications instead of Mono applications. This has been a blocker on fully migrating to .NET Core for years.

Is this something that compressed pointers might help with? They have been used with some degree of success in [Java](https://docs.oracle.com/javase/7/docs/technotes/guides/vm/performance-enhancements-7.html#compressedOop) and [v8](https://docs.google.com/document/d/10qh2-b4C5OtSg-xLwyZpEI5ZihVBPtn1xwKBbQC26yI/edit). Is there something similar planned in .NET Core?

**Comment by weltkante** [comment]
@janvorli mentioned on the linked coreclr issue

> you don't need a pure x86 Ubuntu to build. You can use cross compilation - first build rootfs using the eng/common/cross/build-rootfs.sh script for x86 (run it on x64) and then use it for cross compiling.

So I'll be trying that instead of setting up a x86 VM for building and see if the output happens to be an x86 build. I'll have to note though that how to cross compile via source-build isn't documented at all. I found [someone](https://github.com/dotnet/source-build/issues/113) dropping `./build.sh /p:Platform=arm` in a comment so I'll try it for x86 after doing the rootfs and see if that produces anything, but again: this switch/property isn't documented either for the scripts

Regarding the comment from @omajid 

> Is this something that compressed pointers might help with? They have been used with some degree of success in Java and v8. Is there something similar planned in .NET Core?

Welll, it could probably reduce the pain, but considering that x86 builds are already supposed to work and just aren't documented, I'd prefer getting them working instead of waiting month to years for an implementation of compressed pointers in .NET Core ...

**Comment by dseefeld** [comment]
Thanks for filing this issue.  x86 is not currently active within source-build.  We need to improve our documentation regarding this and bootstrapping.

**Comment by weltkante** [comment]
> x86 is not currently active within source-build

Well, it seemed some people worked on it, as I found #680 and #684 but it has no indication how far they got or whether it was working on some point.

Anyways, reporting back from my attempt of building with rootfs, I tried

- `./cross/build-rootfs.sh x86 xenial`
- `./build.sh /p:Platform=x86`

I'm getting to `Writing package usage data...` which takes very long and fails with out of memory, killing process `dotnet` due to needing around 17 GB memory. I currently don't have a machine with more RAM around so I'm giving up for now. (I don't even know if what I'm doing would result in a valid x86 build in the first place.)

If someone has further hints or knows how to do this I'll try again.

**Comment by ntindle** [comment]
I have a server that it can run on if you can Run what you need. I don’t know how to help but I need this downstream

**Comment by deinok** [comment]
@weltkante If your issue is not having enough RAM, I can build the code for you (with your instructions), I have 32GB of RAM.
I could try to give you SSH access to it.

**Comment by weltkante** [comment]
No, the issue is that the scripts don't work. Cross compiling on an x64 machine against x86 works just fine for the dotnet runtime repo, but building the SDK fails because the scripts were not maintained. It's been a while so maybe things have changed and the issue was not updated.

**Comment by deinok** [comment]
I will try to generate a working corerun and see if it is able to execute something.

**Comment by deinok** [comment]
Then probably we should start executing some CI in the runtime repo for x86, atleast generating a build without running tests

**Comment by weltkante** [comment]
Linking back to dotnet/runtime#7335, my last attempt was [this](https://github.com/dotnet/runtime/issues/7335#issuecomment-629851868) - contains instructions what I did to get it building. Looked straightforward back then. I'll try to rerun my experiments in the next few days.

**Comment by weltkante** [comment]
summary of my tests:
* this repo is removing x86 rootfs from main branch (eng/common/cross) even though all other repos still have it
* this repo seems to have no documentation for cross compiling
* all other repos except dotnet/runtime have no documentation for cross compiling (notably tried dotnet/sdk which has rootfs for x86 but no instructions)
* dotnet/runtime has `--cross --arch x86` parameters to build.sh but they seemed to have regressed

**Comment by janvorli** [comment]
@weltkante the x86 rootfs support is still there and we don't plan to remove it. I was testing that my changes didn't break x86 Linux build using a rootfs generated by the build_rootfs.sh just a week ago. I am not sure where have you seen it removed.

**Comment by weltkante** [comment]
I simply found the x86 subfolder missing from `eng/common/cross` in the main branch, and the `cross` folder in the root is gone, but maybe I'm looking in the wrong places. `eng/common/cross` is where the other repos have their listing of rootfs sources.

**Comment by janvorli** [comment]
It is still here:
https://github.com/dotnet/runtime/tree/main/eng/common/cross/x86

**Comment by weltkante** [comment]
Yes but thats dotnet/runtime and not this repo, the folder is gone from main in this repo (it existed in branches for previous versions). Maybe this repo doesn't need it, but I couldn't get it to work it either way.

**Comment by janvorli** [comment]
Oh, I am sorry, I have not noticed that this issue is in the source-build repo. You can use the eng/common/cross/build_rootfs.sh from the runtime repo to build the rootfs, it is supposed to be the same everywhere. The source one is in the dotnet/arcade repo and it gets automatically copied to the other repos. It seems that the source-build repo doesn't have that copying established.

**Comment by MichaelSimons** [comment]
> the folder is gone from main in this repo

For .NET 6.0, source-build is getting pushed to the individual product repos and will be initiated from the installer repo.  This is WIP and you can read the details at https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build.  This repo will be used for issue tracking and will be used to build the components required by source-build that reside outside of the dotnet organization.

**Comment by weltkante** [comment]
its still unclear how to pass x86 flags to build.sh for this repo, it doesn't seem to take the same parameters as dotnet/runtime, thus this issue for documenting how its supposed to work

for dotnet/runtime the discussion was at dotnet/runtime#7335 as far as I'm aware and it was failing me when I checked last weekend. Maybe I just had a bad commit on top of the main branch, or some details how to build the rootfs changed since I last tried successfully.
